src/ejabberd_c2s.erl

```diff
@@ -1182,6 +1184,13 @@ handle_incoming_message({run_remote_hook, HandlerName, Args}, StateName, StateDa
                        [HandlerName, Args, HandlerState, E]),
             fsm_next_state(StateName,
                            StateData);
+        {update_buffers, NewHandlerState, NewStateData} ->
+            % if a handler sent some stanzas directly it has to pass updated state like this
+            % so that we update stream management buffers
```

Comment:
Stream management is not the only thing to consider here. How about the `csi` buffers when the client using [XEP-0352: CSI](https://xmpp.org/extensions/xep-0352.html) puts the session into inactive?

Reply:
Also, our tests are not ideal, many things are not tested with stream management enabled, not to mention `csi`. I wonder if there is a way to enforce the correct return value indicating the need for updating some parts of the state in case the handler uses the `ejabberd_c2s:send_to_local_user`.

Reply:
Maybe we need to relax the separation here and just allow the handlers to access and modify the full c2s state? The calls to `ejabberd_c2s_state:get_handler_state(HandlerName, StateData)` and `ejabberd_c2s_state:set_handler_state(HandlerName, NewHandlerState, StateData1)` could happen in the handlers. Some of them may not even need their state AFAIR. I think this would be still better than trying to enumerate all possibilities in the case clauses here.

Reply:
I considered returning the whole state, but the design decision was to not do it... But it seems safer indeed, and shouldn't hurt.

Reply:
Right, when making the decision we didn't have in mind delivering to a local user, now considering that I'm in favour of relaxing the separation. Only this will introduce more changes to the entire refactoring at this point, right?